File: photon-core/src/main/java/org/photonvision/vision/estimation/OpenCVHelp.java
package org.photonvision.vision.estimation;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.ejml.simple.SimpleMatrix;
import org.opencv.calib3d.Calib3d;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.MatOfDouble;
import org.opencv.core.MatOfInt;
import org.opencv.core.MatOfPoint;
import org.opencv.core.MatOfPoint2f;
import org.opencv.core.MatOfPoint3f;
import org.opencv.core.Point;
import org.opencv.core.Point3;
import org.opencv.core.RotatedRect;
import org.opencv.imgproc.Imgproc;
import org.photonvision.targeting.TargetCorner;
import org.photonvision.vision.estimation.VisionEstimation.PNPResults;

import edu.wpi.first.cscore.CameraServerCvJNI;
import edu.wpi.first.math.Matrix;
import edu.wpi.first.math.Nat;
import edu.wpi.first.math.Num;
import edu.wpi.first.math.Vector;
import edu.wpi.first.math.geometry.CoordinateSystem;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Transform3d;
import edu.wpi.first.math.geometry.Translation3d;
import edu.wpi.first.math.numbers.*;

public final class OpenCVHelp {

    // static {
    //     try {
    //         CameraServerCvJNI.forceLoad();
    //     } catch (Exception e) {
    //         throw new RuntimeException("Failed to load native libraries!", e);
    //     }
    // }

    public static MatOfDouble matrixToMat(SimpleMatrix matrix) {
        var mat = new Mat(matrix.numRows(), matrix.numCols(), CvType.CV_64F);
        mat.put(0, 0, matrix.getDDRM().getData());
        var wrappedMat = new MatOfDouble();
        mat.convertTo(wrappedMat, CvType.CV_64F);
        mat.release();
        return wrappedMat;
    }
    public static Matrix<Num, Num> matToMatrix(Mat mat) {
        double[] data = new double[(int)mat.total()*mat.channels()];
        var doubleMat = new Mat(mat.rows(), mat.cols(), CvType.CV_64F);
        mat.convertTo(doubleMat, CvType.CV_64F);
        doubleMat.get(0, 0, data);
        return new Matrix<>(new SimpleMatrix(mat.rows(), mat.cols(), true, data));
    }

    /**
     * Creates a new {@link MatOfPoint3f} with these 3d translations.
     * The opencv tvec is a vector with three elements representing {x, y, z} in the
     * EDN coordinate system.
     * @param tvecOutput The tvec to be filled with data
     */
    public static MatOfPoint3f translationToTvec(Translation3d... translations) {
        Point3[] points = new Point3[translations.length];
        for(int i = 0; i < translations.length; i++) {
            var trl = CoordinateSystem.convert(
                translations[i],
                CoordinateSystem.NWU(),
                CoordinateSystem.EDN()
            );
            points[i] = new Point3(trl.getX(), trl.getY(), trl.getZ());
        }
        return new MatOfPoint3f(points);
    }
    /**
     * Returns a new 3d translation from this {@link Mat}.
     * The opencv tvec is a vector with three elements representing {x, y, z} in the
     * EDN coordinate system.
     */
    public static Translation3d tvecToTranslation(Mat tvecInput) {
        float[] data = new float[3];
        var wrapped = new Mat(tvecInput.rows(), tvecInput.cols(), CvType.CV_32F);
        tvecInput.convertTo(wrapped, CvType.CV_32F);
        wrapped.get(0, 0, data);
        wrapped.release();
        return CoordinateSystem.convert(
            new Translation3d(data[0], data[1], data[2]),
            CoordinateSystem.EDN(),
            CoordinateSystem.NWU()
        );
    }

    /**
     * Creates a new {@link MatOfPoint3f} with this 3d rotation.
     * The opencv rvec Mat is a vector with three elements representing the axis
     * scaled by the angle in the EDN coordinate system.
     * (angle = norm, and axis = rvec / norm)
     * @param rvecOutput The rvec Mat to be filled with data
     */
    public static MatOfPoint3f rotationToRvec(Rotation3d rotation) {
        rotation = rotationNWUtoEDN(rotation);
        return new MatOfPoint3f(new Point3(rotation.getQuaternion().toRotationVector().getData()));
    }
    /**
     * Returns a 3d rotation from this {@link Mat}.
     * The opencv rvec Mat is a vector with three elements representing the axis
     * scaled by the angle in the EDN coordinate system.
     * (angle = norm, and axis = rvec / norm)
     */
    public static Rotation3d rvecToRotation(Mat rvecInput) {
        float[] data = new float[3];
        var wrapped = new Mat(rvecInput.rows(), rvecInput.cols(), CvType.CV_32F);
        rvecInput.convertTo(wrapped, CvType.CV_32F);
        wrapped.get(0, 0, data);
        wrapped.release();
        Vector<N3> axis = new Vector<>(Nat.N3());
        axis.set(0, 0, data[0]);
        axis.set(1, 0, data[1]);
        axis.set(2, 0, data[2]);
        return rotationEDNtoNWU(new Rotation3d(axis.div(axis.norm()), axis.norm()));
    }
    
    public static MatOfPoint2f targetCornersToMat(List<TargetCorner> corners) {
        return targetCornersToMat(corners.toArray(TargetCorner[]::new));
    }
    public static MatOfPoint2f targetCornersToMat(TargetCorner... corners) {
        var points = new Point[corners.length];
        for(int i=0; i<corners.length; i++) {
            points[i] = new Point(corners[i].x, corners[i].y);
        }
        return new MatOfPoint2f(points);
    }
    public static TargetCorner[] matToTargetCorners(MatOfPoint2f matInput) {
        var corners = new TargetCorner[(int)matInput.total()];
        float[] data = new float[(int)matInput.total()*matInput.channels()];
        matInput.get(0, 0, data);
        for(int i=0; i<corners.length; i++) {
            corners[i] = new TargetCorner(data[0+2*i], data[1+2*i]);
        }
        return corners;
    }

    /**
     * Reorders the list, optionally indexing backwards and wrapping around to the
     * last element after the first, and shifting all indices in the direction of indexing.
     * 
     * <p>e.g.
     * <p>({1,2,3}, false, 1) -> {2,3,1}
     * <p>({1,2,3}, true, 0) -> {1,3,2}
     * <p>({1,2,3}, true, 1) -> {3,2,1}
     * @param <T>
     * @param elements
     * @param backwards
     * @param shift
     * @return
     */
    public static <T> List<T> reorderCircular(List<T> elements, boolean backwards, int shift) {
        int size = elements.size();
        int dir = backwards ? -1 : 1;
        var reordered = new ArrayList<>(elements);
        for(int i = 0; i < size; i++) {
            int index = (i*dir + shift*dir) % size;
            if(index < 0) index = size + index;
            reordered.set(i, elements.get(index));
        }
        return reordered;
    }

    /**
     * Convert a rotation from EDN to NWU. For example, if you have a rotation X,Y,Z {1, 0, 0}
     * in EDN, this would be XYZ {0, -1, 0} in NWU.
     */
    private static Rotation3d rotationEDNtoNWU(Rotation3d rot) {
        return CoordinateSystem.convert(
            new Rotation3d(),
            CoordinateSystem.NWU(),
            CoordinateSystem.EDN()
        ).plus(
            CoordinateSystem.convert(
                rot,
                CoordinateSystem.EDN(),
                CoordinateSystem.NWU()
            )
        );
    }
    /**
     * Convert a rotation from EDN to NWU. For example, if you have a rotation X,Y,Z {1, 0, 0}
     * in EDN, this would be XYZ {0, -1, 0} in NWU.
     */
    private static Rotation3d rotationNWUtoEDN(Rotation3d rot) {
        return CoordinateSystem.convert(
            new Rotation3d(),
            CoordinateSystem.EDN(),
            CoordinateSystem.NWU()
        ).plus(
            CoordinateSystem.convert(
                rot,
                CoordinateSystem.NWU(),
                CoordinateSystem.EDN()
            )
        );
    }

    /**
     * Project object points from the 3d world into the 2d camera image.
     * The camera properties(intrinsics, distortion) determine the results of
     * this projection.
     * 
     * @param camPose The current camera pose in the 3d world
     * @param camProp The properties of this camera
     * @param objectTranslations The 3d points to be projected
     * @return The 2d points in pixels which correspond to the image of the 3d points on the camera
     */
    public static List<TargetCorner> projectPoints(
            Pose3d camPose, CameraProperties camProp,
            List<Translation3d> objectTranslations) {
        // translate to opencv classes
        var objectPoints = translationToTvec(objectTranslations.toArray(new Translation3d[0]));
        // opencv rvec/tvec describe a change in basis from world to camera
        var basisChange = RotTrlTransform3d.makeRelativeTo(camPose);
        var rvec = rotationToRvec(basisChange.getRotation());
        var tvec = translationToTvec(basisChange.getTranslation());
        var cameraMatrix = matrixToMat(camProp.getIntrinsics().getStorage());
        var distCoeffs = matrixToMat(camProp.getDistCoeffs().getStorage());
        var imagePoints = new MatOfPoint2f();
        // project to 2d
        Calib3d.projectPoints(objectPoints, rvec, tvec, cameraMatrix, distCoeffs, imagePoints);
        
        // turn 2d point Mat into TargetCorners
        var corners = matToTargetCorners(imagePoints);

        // release our Mats from native memory
        objectPoints.release();
        rvec.release();
        tvec.release();
        cameraMatrix.release();
        distCoeffs.release();
        imagePoints.release();

        return Arrays.asList(corners);
    }

    /**
     * Undistort 2d image points using a given camera's intrinsics and distortion.
     * 
     * <p>2d image points from {@link #projectPoints(Pose3d, CameraProperties, Translation3d...)}
     * will naturally be distorted, so this operation is important if the image points
     * need to be directly used (e.g. 2d yaw/pitch).
     * @param camProp The properties of this camera
     * @param corners The distorted image points
     * @return The undistorted image points
     */
    public static List<TargetCorner> undistortPoints(CameraProperties camProp, List<TargetCorner> corners) {
        var points_in = targetCornersToMat(corners);
        var points_out = new MatOfPoint2f();
        var cameraMatrix = matrixToMat(camProp.getIntrinsics().getStorage());
        var distCoeffs = matrixToMat(camProp.getDistCoeffs().getStorage());

        Calib3d.undistortImagePoints(points_in, points_out, cameraMatrix, distCoeffs);
        var corners_out = matToTargetCorners(points_out);

        points_in.release();
        points_out.release();
        cameraMatrix.release();
        distCoeffs.release();

        return Arrays.asList(corners_out);
    }

    /**
     * Get the rectangle with minimum area which bounds this contour. This is useful for finding the center of a bounded
     * contour or the size of the bounding box.
     * 
     * @param corners The corners defining this contour
     * @return Rectangle bounding the contour created by the given corners
     */
    public static RotatedRect getMinAreaRect(List<TargetCorner> corners) {
        var corn = targetCornersToMat(corners);
        var rect = Imgproc.minAreaRect(corn);
        corn.release();
        return rect;
    }
    /**
     * Get the area in pixels of this target's contour. It's important to note that this may
     * be different from the area of the bounding rectangle around the contour.
     * 
     * @param corners The corners defining this contour
     * @return Area in pixels (units of corner x/y)
     */
    public static double getContourAreaPx(List<TargetCorner> corners) {
        var temp = targetCornersToMat(corners);
        var corn = new MatOfPoint(temp.toArray());
        temp.release();
        
        // outputHull gives us indices (of corn) that make a convex hull contour
        var outputHull = new MatOfInt();
        Imgproc.convexHull(corn, outputHull);
        int[] indices = outputHull.toArray();
        outputHull.release();
        var tempPoints = corn.toArray();
        var points = tempPoints.clone();
        for(int i=0; i<indices.length; i++) {
            points[i] = tempPoints[indices[i]];
        }
        corn.fromArray(points);
        // calculate area of the (convex hull) contour
        double area = Imgproc.contourArea(corn);
        corn.release();
        return area;
    }

    /**
     * Finds the transformation(s) that map the camera's pose to the target pose.
     * The camera's pose relative to the target is determined by the supplied
     * 3d points of the target's model and their associated 2d points imaged by the camera.
     * 
     * <p>For planar targets, there may be an alternate solution which is plausible given
     * the 2d image points. This has an associated "ambiguity" which describes the
     * ratio of reprojection error between the "best" and "alternate" solution.
     * 
     * @param camProp The properties of this camera
     * @param modelTrls The translations of the object corners. These should have the object
     *     pose as their origin. These must come in a specific, pose-relative order (in NWU):
     *     <ul>
     *     <li> Point 0: [0, -squareLength / 2,  squareLength / 2]
     *     <li> Point 1: [0,  squareLength / 2,  squareLength / 2]
     *     <li> Point 2: [0,  squareLength / 2, -squareLength / 2]
     *     <li> Point 3: [0, -squareLength / 2, -squareLength / 2]
     *     </ul>
     * @param imageCorners The projection of these 3d object points into the 2d camera image.
     *     The order should match the given object point translations.
     * @return The resulting <b>transformation(s)</b> that map the camera pose to the target pose
     *     and the ambiguity if alternate solutions are also available.
     */
    public static PNPResults solveTagPNP(
            CameraProperties camProp, List<Translation3d> modelTrls, List<TargetCorner> imageCorners) {
        // IPPE_SQUARE expects our corners in a specific order
        modelTrls = reorderCircular(modelTrls, false, 2);
        imageCorners = reorderCircular(imageCorners, false, 2);
        // translate to opencv classes
        var objectPoints = translationToTvec(modelTrls.toArray(new Translation3d[0]));
        var imagePoints = targetCornersToMat(imageCorners);
        var cameraMatrix = matrixToMat(camProp.getIntrinsics().getStorage());
        var distCoeffs = matrixToMat(camProp.getDistCoeffs().getStorage());
        var rvecs = new ArrayList<Mat>();
        var tvecs = new ArrayList<Mat>();
        var rvec = Mat.zeros(3, 1, CvType.CV_32F);
        var tvec = Mat.zeros(3, 1, CvType.CV_32F);
        var reprojectionError = new Mat();
        // calc rvecs/tvecs and associated reprojection error from image points
        Calib3d.solvePnPGeneric(
            objectPoints, imagePoints,
            cameraMatrix, distCoeffs,
            rvecs, tvecs,
            false, Calib3d.SOLVEPNP_IPPE_SQUARE,
            rvec, tvec,
            reprojectionError
        );

        float[] errors = new float[2];
        reprojectionError.get(0, 0, errors);
        // convert to wpilib coordinates
        var best = new Transform3d(
            tvecToTranslation(tvecs.get(0)),
            rvecToRotation(rvecs.get(0))
        );

        var alt = new Transform3d();
        if(tvecs.size() > 1) {
            alt = new Transform3d(
                tvecToTranslation(tvecs.get(1)),
                rvecToRotation(rvecs.get(1))
            );
        }

        // release our Mats from native memory
        objectPoints.release();
        imagePoints.release();
        cameraMatrix.release();
        distCoeffs.release();
        for(var v : rvecs) v.release();
        for(var v : tvecs) v.release();
        rvec.release();
        tvec.release();
        reprojectionError.release();
        return new VisionEstimation.PNPResults(best, alt, errors[0] / errors[1], errors[0], errors[1]);
    }
    /**
     * Finds the transformation that maps the camera's pose to the field origin.
     * The camera's pose relative to the targets is determined by the supplied 3d points of
     * the target's corners on the field and their associated 2d points imaged by the camera.
     * 
     * <p>For planar targets, there may be an alternate solution which is plausible given
     * the 2d image points. This has an associated "ambiguity" which describes the
     * ratio of reprojection error between the "best" and "alternate" solution.
     * 
     * @param camProp The properties of this camera
     * @param objectTrls The translations of the object corners, relative to the field.
     * @param imageCorners The projection of these 3d object points into the 2d camera image.
     *     The order should match the given object point translations.
     * @return The resulting transformation(s) that map the camera pose to the target pose
     *     and the ambiguity if alternate solutions are also available.
     */
    public static PNPResults solveTagsPNP(
            CameraProperties camProp, List<Translation3d> objectTrls, List<TargetCorner> imageCorners) {
        // translate to opencv classes
        var objectPoints = translationToTvec(objectTrls.toArray(new Translation3d[0]));
        var imagePoints = targetCornersToMat(imageCorners);
        var cameraMatrix = matrixToMat(camProp.getIntrinsics().getStorage());
        var distCoeffs = matrixToMat(camProp.getDistCoeffs().getStorage());
        var rvecs = new ArrayList<Mat>();
        var tvecs = new ArrayList<Mat>();
        var rvec = Mat.zeros(3, 1, CvType.CV_32F);
        var tvec = Mat.zeros(3, 1, CvType.CV_32F);
        var reprojectionError = new Mat();
        // calc rvec/tvec from image points
        Calib3d.solvePnPGeneric(
            objectPoints, imagePoints,
            cameraMatrix, distCoeffs,
            rvecs, tvecs,
            false, Calib3d.SOLVEPNP_SQPNP,
            rvec, tvec, reprojectionError
        );

        float[] errors = new float[2];
        reprojectionError.get(0, 0, errors);
        // convert to wpilib coordinates
        var best = new Transform3d(
            tvecToTranslation(tvecs.get(0)),
            rvecToRotation(rvecs.get(0))
        );
        var alt = new Transform3d();
        
        // release our Mats from native memory
        objectPoints.release();
        imagePoints.release();
        cameraMatrix.release();
        distCoeffs.release();
        for(var v : rvecs) v.release();
        for(var v : tvecs) v.release();
        rvec.release();
        tvec.release();
        reprojectionError.release();
        return new PNPResults(best, alt, 0, errors[0], 0);
    }
    public static PNPResults solveTagsPNPRansac(
            CameraProperties camProp, List<Translation3d> objectTrls, List<TargetCorner> imageCorners) {
        // translate to opencv classes
        var objectPoints = translationToTvec(objectTrls.toArray(new Translation3d[0]));
        var imagePoints = targetCornersToMat(imageCorners);
        var cameraMatrix = matrixToMat(camProp.getIntrinsics().getStorage());
        var distCoeffs = matrixToMat(camProp.getDistCoeffs().getStorage());
        var rvec = Mat.zeros(3, 1, CvType.CV_32F);
        var tvec = Mat.zeros(3, 1, CvType.CV_32F);
        var inliers = new Mat();
        // calc rvec/tvec from image points
        Calib3d.solvePnPRansac(
            objectPoints, imagePoints,
            cameraMatrix, distCoeffs,
            rvec, tvec,
            false, 10000,
            1f, 0.99, inliers,
            Calib3d.USAC_MAGSAC
        );
        // System.out.println("--Inliers:");
        // System.out.println(inliers.dump());

        var best = new Transform3d(
            tvecToTranslation(tvec),
            rvecToRotation(rvec)
        );
        var alt = new Transform3d();
        
        // release our Mats from native memory
        objectPoints.release();
        imagePoints.release();
        cameraMatrix.release();
        distCoeffs.release();
        rvec.release();
        tvec.release();
        inliers.release();
        return new PNPResults(best, alt, 0, 0, 0);
    }

    public static RotTrlTransform3d estimateRigidTransform(
            List<Translation3d> measuredTrls, List<Translation3d> knownTrls) {
        var measuredPts = translationToTvec(measuredTrls.toArray(new Translation3d[0]));
        var knownPts = translationToTvec(knownTrls.toArray(new Translation3d[0]));

        var outputMat = Calib3d.estimateAffine3D(measuredPts, knownPts, null, true);
        var outputMatrix = matToMatrix(outputMat);

        // rotation matrix
        var rot = new Rotation3d(outputMatrix.<N3, N3>block(3, 3, 0, 0));
        rot = rotationEDNtoNWU(rot);
        // translation vector
        double[] trlData = outputMatrix.<N3, N1>block(3, 1, 0, 3).getData();
        var trl = new Translation3d(trlData[0], trlData[1], trlData[2]);
        trl = CoordinateSystem.convert(trl, CoordinateSystem.EDN(), CoordinateSystem.NWU());

        var trf = new RotTrlTransform3d(
            rot,
            trl
        );

        // release our Mats from native memory
        measuredPts.release();
        knownPts.release();
        outputMat.release();

        return trf;
    }
}

File: photon-lib/py/photonlibpy/estimation/openCVHelp.py
import math
from typing import Any

import cv2 as cv
import numpy as np
from wpimath.geometry import Rotation3d, Transform3d, Translation3d

from ..targeting import PnpResult, TargetCorner
from .rotTrlTransform3d import RotTrlTransform3d

NWU_TO_EDN = Rotation3d(np.array([[0, -1, 0], [0, 0, -1], [1, 0, 0]]))
EDN_TO_NWU = Rotation3d(np.array([[0, 0, 1], [-1, 0, 0], [0, -1, 0]]))


class OpenCVHelp:
    @staticmethod
    def getMinAreaRect(points: np.ndarray) -> cv.RotatedRect:
        return cv.RotatedRect(*cv.minAreaRect(points))

    @staticmethod
    def translationNWUtoEDN(trl: Translation3d) -> Translation3d:
        return trl.rotateBy(NWU_TO_EDN)

    @staticmethod
    def rotationNWUtoEDN(rot: Rotation3d) -> Rotation3d:
        return -NWU_TO_EDN + (rot + NWU_TO_EDN)

    @staticmethod
    def translationToTVec(translations: list[Translation3d]) -> np.ndarray:
        """Creates a new :class:`np.array` with these 3d translations. The opencv tvec is a vector with
        three elements representing {x, y, z} in the EDN coordinate system.

        :param translations: The translations to convert into a np.array
        """

        retVal: list[list] = []
        for translation in translations:
            trl = OpenCVHelp.translationNWUtoEDN(translation)
            retVal.append([trl.X(), trl.Y(), trl.Z()])
        return np.array(
            retVal,
            dtype=np.float32,
        )

    @staticmethod
    def rotationToRVec(rotation: Rotation3d) -> np.ndarray:
        """Creates a new :class:`.np.array` with this 3d rotation. The opencv rvec Mat is a vector with
        three elements representing the axis scaled by the angle in the EDN coordinate system. (angle =
        norm, and axis = rvec / norm)

        :param rotation: The rotation to convert into a np.array
        """

        retVal: list[np.ndarray] = []
        rot = OpenCVHelp.rotationNWUtoEDN(rotation)
        rotVec = rot.getQuaternion().toRotationVector()
        retVal.append(rotVec)
        return np.array(
            retVal,
            dtype=np.float32,
        )

    @staticmethod
    def avgPoint(points: np.ndarray) -> np.ndarray:
        x = 0.0
        y = 0.0
        for p in points:
            x += p[0, 0]
            y += p[0, 1]
        return np.array([[x / len(points), y / len(points)]])

    @staticmethod
    def pointsToTargetCorners(points: np.ndarray) -> list[TargetCorner]:
        corners = [TargetCorner(p[0, 0], p[0, 1]) for p in points]
        return corners

    @staticmethod
    def cornersToPoints(corners: list[TargetCorner]) -> np.ndarray:
        points = [[[c.x, c.y]] for c in corners]
        return np.array(points)

    @staticmethod
    def projectPoints(
        cameraMatrix: np.ndarray,
        distCoeffs: np.ndarray,
        camRt: RotTrlTransform3d,
        objectTranslations: list[Translation3d],
    ) -> np.ndarray:
        objectPoints = OpenCVHelp.translationToTVec(objectTranslations)
        rvec = OpenCVHelp.rotationToRVec(camRt.getRotation())
        tvec = OpenCVHelp.translationToTVec(
            [
                camRt.getTranslation(),
            ]
        )

        pts, _ = cv.projectPoints(objectPoints, rvec, tvec, cameraMatrix, distCoeffs)
        return pts

    @staticmethod
    def reorderCircular(
        elements: list[Any] | np.ndarray, backwards: bool, shiftStart: int
    ) -> list[Any]:
        """Reorders the list, optionally indexing backwards and wrapping around to the last element after
        the first, and shifting all indices in the direction of indexing.

        e.g.

        ({1,2,3}, false, 1) == {2,3,1}

        ({1,2,3}, true, 0) == {1,3,2}

        ({1,2,3}, true, 1) == {3,2,1}

        :param elements:   list elements
        :param backwards:  If indexing should happen in reverse (0, size-1, size-2, ...)
        :param shiftStart: How much the initial index should be shifted (instead of starting at index 0,
                           start at shiftStart, negated if backwards)

        :returns: Reordered list
        """

        size = len(elements)
        reordered = []
        dir = -1 if backwards else 1
        for i in range(size):
            index = (i * dir + shiftStart * dir) % size
            if index < 0:
                index += size
            reordered.append(elements[index])
        return reordered

    @staticmethod
    def translationEDNToNWU(trl: Translation3d) -> Translation3d:
        """Convert a rotation delta from EDN to NWU. For example, if you have a rotation X,Y,Z {1, 0, 0}
        in EDN, this would be {0, -1, 0} in NWU.
        """

        return trl.rotateBy(EDN_TO_NWU)

    @staticmethod
    def rotationEDNToNWU(rot: Rotation3d) -> Rotation3d:
        """Convert a rotation delta from NWU to EDN. For example, if you have a rotation X,Y,Z {1, 0, 0}
        in NWU, this would be {0, 0, 1} in EDN.
        """

        return -EDN_TO_NWU + (rot + EDN_TO_NWU)

    @staticmethod
    def tVecToTranslation(tvecInput: np.ndarray) -> Translation3d:
        """Returns a new 3d translation from this :class:`.Mat`. The opencv tvec is a vector with three
        elements representing {x, y, z} in the EDN coordinate system.

        :param tvecInput: The tvec to create a Translation3d from
        """

        return OpenCVHelp.translationEDNToNWU(Translation3d(tvecInput))

    @staticmethod
    def rVecToRotation(rvecInput: np.ndarray) -> Rotation3d:
        """Returns a 3d rotation from this :class:`.Mat`. The opencv rvec Mat is a vector with three
        elements representing the axis scaled by the angle in the EDN coordinate system. (angle = norm,
        and axis = rvec / norm)

        :param rvecInput: The rvec to create a Rotation3d from
        """

        return OpenCVHelp.rotationEDNToNWU(Rotation3d(rvecInput))

    @staticmethod
    def solvePNP_Square(
        cameraMatrix: np.ndarray,
        distCoeffs: np.ndarray,
        modelTrls: list[Translation3d],
        imagePoints: np.ndarray,
    ) -> PnpResult | None:
        """Finds the transformation(s) that map the camera's pose to the target's pose. The camera's pose
        relative to the target is determined by the supplied 3d points of the target's model and their
        associated 2d points imaged by the camera. The supplied model translations must be relative to
        the target's pose.

        For planar targets, there may be an alternate solution which is plausible given the 2d image
        points. This has an associated "ambiguity" which describes the ratio of reprojection error
        between the "best" and "alternate" solution.

        This method is intended for use with individual AprilTags, and will not work unless 4 points
        are provided.

        :param cameraMatrix: The camera intrinsics matrix in standard opencv form
        :param distCoeffs:   The camera distortion matrix in standard opencv form
        :param modelTrls:    The translations of the object corners. These should have the object pose as
                             their origin. These must come in a specific, pose-relative order (in NWU):

                             - Point 0: [0, -squareLength / 2, squareLength / 2]
                             - Point 1: [0, squareLength / 2, squareLength / 2]
                             - Point 2: [0, squareLength / 2, -squareLength / 2]
                             - Point 3: [0, -squareLength / 2, -squareLength / 2]
        :param imagePoints:  The projection of these 3d object points into the 2d camera image. The order
                             should match the given object point translations.

        :returns: The resulting transformation that maps the camera pose to the target pose and the
                  ambiguity if an alternate solution is available.
        """
        modelTrls = OpenCVHelp.reorderCircular(modelTrls, True, -1)
        imagePoints = np.array(OpenCVHelp.reorderCircular(imagePoints, True, -1))
        objectMat = np.array(OpenCVHelp.translationToTVec(modelTrls))

        alt: Transform3d | None = None
        reprojectionError: cv.typing.MatLike | None = None
        best: Transform3d = Transform3d()

        for tries in range(2):
            # calc rvecs/tvecs and associated reprojection error from image points
            retval, rvecs, tvecs, reprojectionError = cv.solvePnPGeneric(
                objectMat,
                imagePoints,
                cameraMatrix,
                distCoeffs,
                flags=cv.SOLVEPNP_IPPE_SQUARE,
            )

            # convert to wpilib coordinates
            best = Transform3d(
                OpenCVHelp.tVecToTranslation(tvecs[0]),
                OpenCVHelp.rVecToRotation(rvecs[0]),
            )
            if len(tvecs) > 1:
                alt = Transform3d(
                    OpenCVHelp.tVecToTranslation(tvecs[1]),
                    OpenCVHelp.rVecToRotation(rvecs[1]),
                )

            # check if we got a NaN result
            if reprojectionError is not None and not math.isnan(
                reprojectionError[0, 0]
            ):
                break
            else:
                pt = imagePoints[0]
                pt[0, 0] -= 0.001
                pt[0, 1] -= 0.001
                imagePoints[0] = pt

        # solvePnP failed
        if reprojectionError is None or math.isnan(reprojectionError[0, 0]):
            print("SolvePNP_Square failed!")
            return None

        if alt:
            return PnpResult(
                best=best,
                bestReprojErr=reprojectionError[0, 0],
                alt=alt,
                altReprojErr=reprojectionError[1, 0],
                ambiguity=reprojectionError[0, 0] / reprojectionError[1, 0],
            )
        else:
            # We have no alternative so set it to best as well
            return PnpResult(
                best=best,
                bestReprojErr=reprojectionError[0],
                alt=best,
                altReprojErr=reprojectionError[0],
            )

    @staticmethod
    def solvePNP_SQPNP(
        cameraMatrix: np.ndarray,
        distCoeffs: np.ndarray,
        modelTrls: list[Translation3d],
        imagePoints: np.ndarray,
    ) -> PnpResult | None:
        """Finds the transformation that maps the camera's pose to the origin of the supplied object. An
        "object" is simply a set of known 3d translations that correspond to the given 2d points. If,
        for example, the object translations are given relative to close-right corner of the blue
        alliance(the default origin), a camera-to-origin transformation is returned. If the
        translations are relative to a target's pose, a camera-to-target transformation is returned.

        There must be at least 3 points to use this method. This does not return an alternate
        solution-- if you are intending to use solvePNP on a single AprilTag, see {@link
        #solvePNP_SQUARE} instead.

        :param cameraMatrix: The camera intrinsics matrix in standard opencv form
        :param distCoeffs:   The camera distortion matrix in standard opencv form
        :param objectTrls:   The translations of the object corners, relative to the field.
        :param imagePoints:  The projection of these 3d object points into the 2d camera image. The order
                             should match the given object point translations.

        :returns: The resulting transformation that maps the camera pose to the target pose. If the 3d
                  model points are supplied relative to the origin, this transformation brings the camera to
                  the origin.
        """

        objectMat = np.array(OpenCVHelp.translationToTVec(modelTrls))

        retval, rvecs, tvecs, reprojectionError = cv.solvePnPGeneric(
            objectMat, imagePoints, cameraMatrix, distCoeffs, flags=cv.SOLVEPNP_SQPNP
        )

        error = reprojectionError[0, 0]
        best = Transform3d(
            OpenCVHelp.tVecToTranslation(tvecs[0]), OpenCVHelp.rVecToRotation(rvecs[0])
        )

        if math.isnan(error):
            return None

        # We have no alternative so set it to best as well
        result = PnpResult(best=best, bestReprojErr=error, alt=best, altReprojErr=error)
        return result
